noxfiles/git_nox.py

```diff
@@ -0,0 +1,154 @@
+    - (default) = don't actually apply the tag or push, just show the tag that will be generated
+    - only_tag = generate and apply the a tag locally to the current commit, but don't push the tag
+    - push = generate, apply and push a tag
+    """
+    from git.repo import Repo
+
+    posargs_set = set(session.posargs)
+    repo = Repo()
+    all_tags = get_all_tags(repo)
+
+    generated_tag = generate_tag(
+        session, repo.active_branch.name, all_tags
+    )  # generate a tag based on the current repo state
+
+    # if no args are passed, it's a dry run
+    if ONLY_TAG not in posargs_set and PUSH not in posargs_set:
+        session.log(f"Dry-run -- would generate tag: {generated_tag}")
+        return
+
+    session.log(f"Tagging current HEAD commit with tag: {generated_tag}")
+    repo.create_tag(generated_tag)
+
+    if PUSH in posargs_set and ONLY_TAG not in posargs_set:
+        # push the tag if we've been told to
+        session.log(f"Pushing tag {generated_tag} to remote (origin)")
+        repo.remotes.origin.push(generated_tag)
+
+
+def next_release_increment(session: nox.Session, all_tags: List):
+    """Helper to generate the next release 'increment' based on latest release tag found"""
+    latest_release = next(
+        tag.name for tag in all_tags if re.fullmatch(RELEASE_TAG_REGEX, tag.name)
+    )
+    if not latest_release:  # this would be bad...
+        session.error("Could not identify the latest release!")
+    latest_release = Version(latest_release)
+    return Version(
+        f"{latest_release.major}.{latest_release.minor}.{latest_release.micro + 1}"
+    )
+
+
+def increment_tag(
+    session: nox.Session,
+    all_tags: List,
+    version_number: str,
+    tag_type: TagType,
+):
+    """
+    Utility method to generate the "next" appropriate tag on the given version number
+    with the given tag type.
+    """
+    version_branch_tag_pattern = VERSION_TAG_REGEX.format(
+        version=version_number, tag_type=tag_type.value
+    )
+    # find our latest existing tag for this version/type
+    latest_tag = next(
+        (
+            re.fullmatch(version_branch_tag_pattern, tag.name)
+            for tag in all_tags
+            if re.fullmatch(version_branch_tag_pattern, tag.name)
+        ),
+        None,
+    )
+    if latest_tag:  # if we have an existing tag for this version/type, increment it
+        session.log(
+            f"Found existing {tag_type.name.lower()} tag {latest_tag.group(0)}, incrementing it"
+        )
+        tag_increment = int(latest_tag.group(1)) + TAG_INCREMENT  # increment the tag
+        return f"{version_number}{tag_type.value}{tag_increment}"
+        # return the full {version_number}{tag_type}{increment}
+
+    # we don't have an existing tag for this version/type, so start at our initial increment
+    return f"{version_number}{tag_type.value}{INITIAL_TAG_INCREMENT}"
+
+
+def generate_tag(session: nox.Session, branch_name: str, all_tags: List) -> str:
```

Comment:
nit: I think I'd like to see some `session.log(...)` lines in here explaining the magic. Something like:
```
Fetching current tag and branch to generate a new tag: tag = 2.10.0, branch = 'main'
Current branch 'main' matched main branch, generating a new BETA tag...
Generated new tag: 2.10.0b1
```

Reply:
added some extra logging in https://github.com/ethyca/fides/pull/2979/commits/32bd7487b0e8b117aa2d298ce0afc5c5514cb588, resolving